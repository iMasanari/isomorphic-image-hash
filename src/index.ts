import { dhashFromImageData } from './hash/dhash'
import { getImageDataForBrowser } from './image-data/browser'
import { getGifImageData } from './image-data/gif'
import { getJpegImageData } from './image-data/jpeg'
import { getPngImageData } from './image-data/png'

export { distance } from './utils/distance'
export { dhashFromImageData }

export interface ImageDataLike {
  width: number
  height: number
  data: Uint8Array | Uint8ClampedArray
}

export const getImageData = (path: string): Promise<ImageDataLike> => {
  // for browser
  if (typeof document === 'object') {
    return getImageDataForBrowser(path)
  }

  if (path.endsWith('.png')) {
    return getPngImageData(path)
  }

  if (path.endsWith('.gif')) {
    return getGifImageData(path)
  }

  // JPEG
  return getJpegImageData(path)
}

export const dhash = async (image: string | ImageDataLike) => {
  const imageData = typeof image === 'string' ? await getImageData(image) : image

  return dhashFromImageData(imageData)
}
