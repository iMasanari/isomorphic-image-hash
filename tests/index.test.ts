import Jimp from 'jimp'
import { describe, expect, it } from 'vitest'
import { dhash, dhashFromImageData, distance } from '../src'

const MANDRILL_JPEG_HASH = '0001111110101011111010100110100001101000001110000101001001101000'

describe('dhash', () => {
  it('read jpeg', async () => {
    const hash = await dhash(`${__dirname}/images/mandrill.jpg`)

    expect(hash).toBe(MANDRILL_JPEG_HASH)
  })

  it('read png', async () => {
    const hash = await dhash(`${__dirname}/images/mandrill.png`)
    const diff = distance(hash, MANDRILL_JPEG_HASH)

    expect(diff).toBeLessThanOrEqual(10)
    expect(hash).matchSnapshot()
  })

  it('read gif', async () => {
    const hash = await dhash(`${__dirname}/images/mandrill.gif`)
    const diff = distance(hash, MANDRILL_JPEG_HASH)

    expect(diff).toBeLessThanOrEqual(10)
    expect(hash).matchSnapshot()
  })

  it('read imageData', async () => {
    const { bitmap } = await Jimp.read(`${__dirname}/images/mandrill.jpg`)

    const hash = await dhash({
      width: bitmap.width,
      height: bitmap.height,
      data: new Uint8ClampedArray(bitmap.data),
    })

    expect(hash).toBe(MANDRILL_JPEG_HASH)
  })
})

describe('dhashFromImageData', () => {
  it('read imageData', async () => {
    const { bitmap } = await Jimp.read(`${__dirname}/images/mandrill.jpg`)

    const hash = dhashFromImageData({
      width: bitmap.width,
      height: bitmap.height,
      data: new Uint8ClampedArray(bitmap.data),
    })

    expect(hash).toBe(MANDRILL_JPEG_HASH)
  })
})
