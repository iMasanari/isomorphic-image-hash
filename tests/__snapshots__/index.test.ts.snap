// Vitest Snapshot v1

exports[`dhash > read gif 1`] = `"0001111110101011111010100110100001101000001110000100011001101000"`;

exports[`dhash > read png 1`] = `"0001111110101011111010100110100001101000001110000101001001101000"`;
